banking/ebics/utils.py

```diff
@@ -332,6 +336,88 @@ def upload_camt_file():
 	process_camt_document(camt_document, bank_account)
 
 
+@frappe.whitelist()
+def upload_mt940_file():
+	frappe.has_permission("Bank Transaction", "create", throw=True)
+
+	file_bytes = frappe.local.uploaded_file
+	bank_account = frappe.form_dict.docname
+
+	register_fintech()
+
+	from fintech.swift import parse_mt940
+
+	mt940_data = file_bytes.decode()
+	statements: list[MT940Statement] = parse_mt940(mt940_data)
+
+	for statement in statements:
+		process_mt940_statement(statement, bank_account)
+
+
+def process_mt940_statement(statement: MT940Statement, bank_account: str):
+	"""Process a single MT940 statement and create bank transactions"""
+
+	company = frappe.db.get_value("Bank Account", bank_account, "company")
+	currency = statement["balance_open"]["currency"]
+	for transaction in statement["transactions"]:
+		create_mt940_bank_transaction(bank_account, company, transaction, currency)
+
+
+def create_mt940_bank_transaction(
+	bank_account: str, company: str, transaction: MT940Transaction, currency: str
+):
+	"""Create a bank transaction from MT940 transaction data"""
+
+	amount = transaction["amount"] or 0
+	description = (
+		transaction["description"]
+		or transaction.get("sepa", {}).get("SVWZ")
+		or " ".join(transaction.get("purpose", []))
+	)
+	ref = (
+		transaction.get("sepa", {}).get("EREF")
+		or transaction["reference"]
+		or transaction["bank_reference"]
+		or ""
+	)
+
+	values_to_hash = [
+		transaction["date"],
+		transaction["account"],
+		transaction["name"],
+		transaction["amount"],
+		transaction["reference"],
+		transaction["iban"],
+	]
```

Comment:
It doesn't make sense to use `transaction["iban"]` for generating the hash, when what we're really interested in is `transaction["iban"] or transaction["account"] or ""`. The same goes for reference and party name.

Reply:
I also did this the wrong way for SEPA transactions. But now it's too late to change – once we change the hashing 'algorithm', the existing data will no longer match newly generated hashes. So, considering historical data, the ideal solution would be that the same transaction results in the same hash, regardless of if it was imported via SEPA or MT940.